src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -132,23 +156,73 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
         /// </summary>
         private BoundDeconstructionDeconstructStep MakeNonTupleDeconstructStep(
                                                             BoundDeconstructValuePlaceholder targetPlaceholder,
-                                                            AssignmentExpressionSyntax syntax,
+                                                            ExpressionSyntax syntax,
                                                             DiagnosticBag diagnostics,
-                                                            ImmutableArray<DeconstructionVariable> variables,
+                                                            ArrayBuilder<DeconstructionVariable> variables,
                                                             ArrayBuilder<BoundDeconstructionDeconstructStep> deconstructionSteps,
                                                             ArrayBuilder<BoundDeconstructionAssignmentStep> assignmentSteps)
         {
             // symbol and parameters for Deconstruct
             ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders;
-            var deconstructInvocation = MakeDeconstructInvocationExpression(variables.Length, targetPlaceholder, syntax, diagnostics, out outPlaceholders);
-
+            var deconstructInvocation = MakeDeconstructInvocationExpression(variables.Count, targetPlaceholder, syntax, diagnostics, out outPlaceholders);
             if (deconstructInvocation.HasAnyErrors)
             {
                 return null;
             }
-            else
+
+            SetInferredTypes(variables, outPlaceholders.SelectAsArray(p => p.Type));
+
+            return new BoundDeconstructionDeconstructStep(syntax, deconstructInvocation, targetPlaceholder, outPlaceholders);
+        }
+
+        /// <summary>
+        /// Inform the variables about found types (whether one was found or not).
+        /// </summary>
+        static private void SetInferredTypes(ArrayBuilder<DeconstructionVariable> variables, ImmutableArray<TypeSymbol> foundTypes)
+        {
+            if (!variables.Any(v => IsDeconstructionLocalPendingInference(v)))
+            {
+                return;
+            }
+
+            var matchCount = Math.Min(variables.Count, foundTypes.Length);
+            for (int i = 0; i < matchCount; i++)
             {
-                return new BoundDeconstructionDeconstructStep(syntax, deconstructInvocation, targetPlaceholder, outPlaceholders);
+                var variable = variables[i];
+                if (IsDeconstructionLocalPendingInference(variable))
+                {
+                    var local = ((DeconstructionLocalPendingInference)variable.Single).SetInferredType(foundTypes[i], true);
```

Comment:
Please spell out the type.

---
In reply to: [68801612](https://github.com/dotnet/roslyn/pull/12224#discussion_r68801612) [](ancestors = 68801612)